src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,22 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
     ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
     ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
     ;; set with respect to the alpha roots for any other primary type T2, we know that we will
     ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    ;; Therefore we can reference equality in the interest of maximizing performance and fallback to equivalence,
+    ;; as fact-types can be defined by consumers and thus might not be identical.
+    (let [other ^AlphaRootsWrapper other]
+      (and (= fact-type-hash (.fact-type-hash other))
```

Comment:
- I don't see any scenarios where the hash field would be nil.  We're eagerly population it at the time the AlphaRootsWrapper is created.
- Using == seems reasonable, that should have less overhead.
- Hasheq returns a primitive int: https://github.com/clojure/clojure/blob/clojure-1.8.0/src/jvm/clojure/lang/Util.java#L164
- I think that the call to the Clojure hash function will box the int to an Integer since it doesn't have any type hints preventing that, but I could be wrong.  https://github.com/clojure/clojure/blob/clojure-1.8.0/src/clj/clojure/core.clj#L4961

I agree with changing to ==.  At first glance it seems like the primitive type hints you suggest would be incorrect though.